datasets.py

```diff
@@ -31,47 +38,188 @@ def create_mnist_datasets(
+    if audio.dtype.is_integer:
+        # detected clipping using just tf.int16.max to normalize...
+        audio = tf.cast(audio, tf.float32) / (tf.int16.max + 1.0)
+        
+    # remove DC offset
+    audio_mean = tf.math.reduce_mean(audio, axis=0) 
+    audio -= audio_mean
+    
+    if with_labels:
+        return (audio, example['label'])
+    return audio
+
+
+def _load_spoken_digit_dataset_tf(percent_load, load_labels, download=False):
+    # this dataset only has 'train' split
+    all_data = tfds.load(name='spoken_digit',
+                         split=[f'train[:{percent_load}%]'],
+                         shuffle_files=False,
+                         as_supervised=False,
+                         download=download)
+
+    all_data = all_data[0]
+    all_data = all_data.map(lambda x: _preprocess_spoken_digit_audio_tf(x, load_labels), 
+                            num_parallel_calls=tf.data.AUTOTUNE,
+                            deterministic=False)
+    return all_data
+
+
+def convert_raw_spoken_digit_dataset(destination_dir, percent_load=100, download=False):
+    if not destination_dir.is_dir():
+        assert(destination_dir.parent.is_dir())
+        Path.mkdir(destination_dir)
+
+    all_data = _load_spoken_digit_dataset_tf(percent_load, True, download=download)
+    total_count = len(all_data)
+    print(f'Writing {total_count} examples as numpy arrays')
+    
+    example_id = 0
+    for example_audio, example_label in all_data:
+        destination_path = destination_dir / f'example_{example_id}.npy'
+        audio = example_audio.numpy()
+        label = int(example_label)
+        example_type = np.dtype([('label', np.int8), ('audio', np.float32, (len(audio),))])
+        example_array = np.rec.array((label, audio), dtype=example_type)
+        np.save(destination_path, example_array)
+        example_id += 1
+
+def _get_spoken_digit_feature_channels(feature_name):
+    return 2 if feature_name == 'spectrum_norm' else 1
+
+def _get_spoken_digit_signature_and_padded_shape(config, load_labels):
+    windowed_frame_config = config['dsp']['windowed_frames']
+    frame_size = dsp.get_windowed_frame_size(windowed_frame_config)
+    num_bins = frame_size // 2 + 1
+    
+    stft_features_config = config['dsp']['stft_features']
+    feature_names = stft_features_config['features']
+    
+    signature = {}
+    padded_shapes = {}
+    for feature_name in feature_names:
+        num_channels = _get_spoken_digit_feature_channels(feature_name)
+        signature[feature_name] = tf.TensorSpec(shape=(None, num_bins, num_channels), dtype=tf.float32)
+        padded_shapes[feature_name] = (None, num_bins, num_channels)
+        
+    signature = [signature]
+        
+    if load_labels:
+        signature.append(tf.TensorSpec(shape=(), dtype=tf.int8))
+        padded_shapes = (padded_shapes, ())
+    
+    return tuple(signature), padded_shapes
+
+
+def _extract_spoken_digit_features_npy(example, config, load_labels):
```

Comment:
For this and other functions below, maybe except for the top level one: generally not a fan of passing a generic `config` while the function makes assumptions about what it expects to exist in the config. Each function seems to use 2 or 3 specific values, so I'd just pass those as explicit parameters.

Reply:
The reason I went for the generic config approach is that depending on the kinds of features we end up using we would have a lot of individual parameters and the API could change a lot. Doing it this way makes it easier to accommodate changes later without changing the API. 

In general I agree with the point you're trying to make, and in most situations I would agree with you, but this does streamline things a lot. Maybe we ought to have some sort of 'config validation' or a README or something if you're worried about it being too opaque. 

My though was that since there's this large collection of parameters, many of which are shared between the various functions, it's convenient to gradually build up all the necessary information and collect it in this object. We might start with a couple explicit arguments like 'window_type' and 'window_size' but from there a lot of stuff is computed automatically during the various steps from 'raw audio' to 'stft feature frames' and though I agree that it's confusing right now because it's not documented, letting those functions responsible for each step handle inserting the required information into the config object avoids a lot of clutter and forwarding along a bunch of values that kind of all need to be bundled together anyway.

At the moment it might be manageable to pass everything individually but again, this approach makes it easier to add / remove functionality as we try things without having to continuously change the API as we go.